### PR TITLE
Addon-docs: Jest MDX transform for storyshots

### PR DESCRIPTION
**Pull request by ndelangen**
Issue: #7223

This replaces https://github.com/storybookjs/storybook/pull/7330 since that PR no longer has a repository and cannot be contributed on anymore :(

## What I did

Wrote a jest transform for mdx files.

## How to test

add the following to your jest config:
```json
{
  "transform": {
    "^.+\\.[tj]sx?$": "babel-jest",
    "^.+\\.mdx?$": "@storybook/addon-docs/jest-transform-mdx",
  }
}
```

### CODE REVIEW

**Comment by edelgado** [comment]
Can't wait for this either, thanks! Maybe remove the dependency on `deasync-promise` as explained in https://github.com/storybookjs/storybook/pull/7330#issuecomment-532885357 ? Thanks!

**Comment by ndelangen** [comment]
Would you like to contribute to this PR @edelgado?

**Comment by edelgado** [comment]
@ndelangen sure thing! Thanks for the access, will submit a commit soon.

**Review by shilman — CHANGES_REQUESTED** [review]
NICE!! looking very clean. would it be possible to update the documentation as well?

**Comment by edelgado** [comment]
@shilman Absolutely! Added to both add-ons to make it easier on the users.

**Review by shilman — CHANGES_REQUESTED** [review]
Thanks @edelgado, this looks great. Last request: can we add this to our monorepo somehow to test it?

**Comment by edelgado** [comment]
> Thanks @edelgado, this looks great. Last request: can we add this to our monorepo somehow to test it?

Hi @shilman, not sure what you mean by this.  First time contributing so I'm not familiar with that task. Could you point me in the right direction please?

**Comment by ndelangen** [comment]
@shilman what do you require?

File: addons/docs/README.md [hunk 1]
```diff
@@ -121,6 +121,20 @@ configure(require.context('../src', true, /\.stories\.(js|mdx)$/), module);
...
+If using in conjunction with the [storyshots add-on](../storyshots/storyshots-core/README.md), you will need to 
+configure Jest to transform MDX stories into something Storyshots can understand:
+                                                                                               
+Add the following to your Jest configuration:
+
+```json
+{
+ "transform": {
+   "^.+\\.[tj]sx?$": "babel-jest",
+   "^.+\\.mdx?$": "@storybook/addon-docs/jest-transform-mdx"
```
Comment (Hypnosphi) [inline addons/docs/README.md]: Does it mean that we load `.md` extension as well?

**Comment by shilman** [comment]
@Hypnosphi I tried adding MDX snapshotting to the HTML example and am getting this error:

```
$ yarn jest --testPathPattern htmlshots.test.js
yarn run v1.19.0
$ /Users/shilman/projects/storybookjs/storybook/node_modules/.bin/jest --testPathPattern htmlshots.test.js
 FAIL  examples/html-kitchen-sink/tests/htmlshots.test.js
  ● Test suite failed to run

    SyntaxError: unknown: Expected corresponding JSX closing tag for <MDXLayout> (30:40)

      28 | <p>{\`Hallelujah! HTML is working out of the box without modification.
      29 | `}</code></pre>
    > 30 | <p>{`How you like them apples?!\`}`}</p></p>
         |                                         ^
      31 | <pre><code parentName="pre" {...{}}>{`<Story name="heading" height="100px" mdxType="Story">
      32 | `}</code></pre>
      33 | <p>{`  {'`}<h1>{`Hello World`}</h1>{`'}`}</p>
```

Flying blind here...

**Comment by Hypnosphi** [comment]
@shilman looks like you're adding the transform twice, as it's already present in `config.transform`.

In my case, the html snapshotting went [OK](https://github.com/storybookjs/storybook/pull/8189/files#diff-0e77f87cfc13fb718f853594cd0233e5), but storyshots-integrity complained, probably because its algorythm doesn't know anything about mdx extension yet

**Comment by shilman** [comment]
@Hypnosphi I fixed the integrity error in my latest commit. However, I'm not sure what's going on with Vue/Angular.

**Comment by Hypnosphi** [comment]
@shilman not really

**Comment by shilman** [comment]
@ndelangen and I were able to get angular working. Vue is not working due to a babel issue, and I've temporarily disabled it to move this forward. I'll follow up with @Aaron-Pool to see if we can get that working. Pretty sure it's more of a Vue/Babel issue than a Jest/MDX one.

File: addons/docs/README.md [hunk 2]
```diff
@@ -121,6 +121,20 @@ configure(require.context('../src', true, /\.stories\.(js|mdx)$/), module);
...
 
+If using in conjunction with the [storyshots add-on](../storyshots/storyshots-core/README.md), you will need to 
+configure Jest to transform MDX stories into something Storyshots can understand:
+                                                                                               
+Add the following to your Jest configuration:
+
+```json
+{
+ "transform": {
+   "^.+\\.[tj]sx?$": "babel-jest",
```
Comment (albert-schilling) [inline addons/docs/README.md:132]: Isn't there a pipe symbol missing: [t|j] ? As here: https://github.com/facebook/jest/tree/master/packages/babel-jest#setup

File: addons/storyshots/storyshots-core/README.md [hunk 3]
```diff
@@ -178,6 +178,23 @@ StoryShots addon for Preact is dependent on [preact-render-to-json](https://gith
...
+If using the [Docs add-on](../../docs/README.md) with 
+[MDX stories](../../docs/docs/mdx.md) you will need
+to configure Jest to transform MDX stories into something Storyshots can understand:
+
+Add the following to your Jest configuration:
+
+```json
+{
+  "transform": {
+    "^.+\\.[tj]sx?$": "babel-jest",
```
Comment (albert-schilling) [inline addons/storyshots/storyshots-core/README.md:192]: Isn't the pipe symbol necessary? [t|j]

**Pull request merged**